exonum/src/api/node/public/explorer.rs

```diff
@@ -276,6 +308,72 @@ impl ExplorerApi {
             })
     }
 
+    /// Return call status of committed transaction.
+    pub fn call_status(
+        snapshot: &dyn Snapshot,
+        query: CallStatusQuery,
+    ) -> Result<HashMap<CallKind, CallStatusInfo>, ApiError> {
+        let explorer = BlockchainExplorer::new(snapshot);
+
+        let tx_info = explorer.transaction(&query.hash).ok_or_else(|| {
+            let description = serde_json::to_string(&json!({ "type": "unknown" })).unwrap();
+            ApiError::NotFound(description)
+        })?;
+        let tx_info = match tx_info {
+            TransactionInfo::Committed(info) => Ok(info),
+            TransactionInfo::InPool { .. } => {
+                let description =
+                    serde_json::to_string(&json!({ "type": "not committed" })).unwrap();
+                Err(ApiError::NotFound(description))
+            }
+        }?;
+
+        let block_height = tx_info.location().block_height();
+        let block_info = explorer.block(block_height).ok_or_else(|| {
+            ApiError::NotFound(format!(
+                "Requested block height ({}) exceeds the blockchain height ({})",
+                tx_info.location().block_height(),
+                explorer.height()
+            ))
+        })?;
+
+        let mut call_statuses = HashMap::new();
+        let mut fill_call_status = |call_in_block, call_type| {
+            let status = explorer.call_status(block_height, &call_in_block);
+            let call_proof = block_info.error_proof(call_in_block);
+
+            call_statuses.insert(call_type, CallStatusInfo { status, call_proof });
+        };
+        if let Some(call_type) = query.call_kind {
+            let call_in_block = match call_type {
+                CallKind::BeforeTransactions => CallInBlock::before_transactions(
+                    tx_info.content().payload().call_info.instance_id,
+                ),
+                CallKind::Transaction => {
+                    CallInBlock::transaction(tx_info.location().position_in_block())
+                }
+                CallKind::AfterTransactions => CallInBlock::after_transactions(
+                    tx_info.content().payload().call_info.instance_id,
+                ),
+            };
+            fill_call_status(call_in_block, call_type);
+        } else {
+            let call_in_block: CallInBlock =
+                CallInBlock::before_transactions(tx_info.content().payload().call_info.instance_id);
+            fill_call_status(call_in_block, CallKind::BeforeTransactions);
+
+            let call_in_block: CallInBlock =
+                CallInBlock::transaction(tx_info.location().position_in_block());
+            fill_call_status(call_in_block, CallKind::Transaction);
+
+            let call_in_block: CallInBlock =
+                CallInBlock::after_transactions(tx_info.content().payload().call_info.instance_id);
+            fill_call_status(call_in_block, CallKind::AfterTransactions);
```

Comment:
This code looks unreadalbe, here we need to decomposite this method into the set of clear helpers.

Reply:
And boilerplate code should be generalized if possible.